rubocop: generate_completions DSL

max-ae opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
related to #13536 

needs https://github.com/Homebrew/homebrew-core/pull/105707

cc @Rylan12 

I have achieved to implement a preliminary RuboCop, but I will appreciate any feedback, since I'm sure there's room for improvement.

The Cop is able to correct
```
(bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
```
but I have been unable to come up with a solution to get and replace with all shells whilst only walking the AST once.

The Cop also warns for the 
```
output = Utils.safe_popen_read(bin/"foo", "completion", "bash")
(bash_completion/"foo").write output
```
pattern, but has no corrector, as discussed in #13536.

Please feel free to push any commits improving my implementation 😄

Rylan12 commented:
> but I have been unable to come up with a solution to get and replace with all shells whilst only walking the AST once.

Not sure, I guess I haven't tried. I think you could probably write a pattern that checks for multiple instances right after each other. That would probably be pretty easy to implement. You also could probably do something like this (not tested at all):

```ruby
def audit_formula(_node, _class_node, _parent_class_node, body_node)
  install = find_method_def(body_node, :install)

  completion_nodes_to_fix = []

  correctable_shell_completion_node(install) do |node, shell, base_name, binary, cmd, shell_parameter|
    ...

    # Instead of calling `offending_node` and `problem`:
    completion_nodes_to_fix << [node, replacement]
  end

  completion_nodes_to_fix[0...-1].each do |node, _|
    offending_node(node)
    # The wording here is terrible but is just an example
    problem "Use a single `generate_completions` call to generate completions" do |corrector|
      corrector.replace(@offensive_node.source_range, "")
    end
  end

  # Create the single `generate_completions` call
  offending_node(completion_nodes_to_fix.last.first)
  problem "Use `#{replacement}` instead of `#{@offensive_node.source}`." do |corrector|
    corrector.replace(@offensive_node.source_range, replacement)
  end
end
```

That's just an example, I haven't thought through the logic very thoroughly so there might be a better way to do it.

If those don't work, you could probably write a separate rubocop that combines all of the individual `generate_completions` calls into one combined call.

MikeMcQuaid reviewed (commented):
Looks good so far! Unit tests for this should make it much easier to test.

Rylan12 commented:
> Looks good so far! Unit tests for this should make it much easier to test.

Ah yes, I forgot to mention tests in my initial comment. RuboCop tests are, luckily, pretty easy to make. I'd recommend looking at the [`ShellCommands` test](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/rubocops/shell_commands_spec.rb) since it will be almost identical to what you need here. Happy to give more guidance if needed!

max-ae commented:
Thank you for the feedback and helpful example code! I'll update the PR with the proposed changes as soon as I refactored the DSL PR, and will also include Unit Tests when doing so.

max-ae commented:
Hello there! I have updated the PR to reflect the changes made in the DSL PR, and also included some other fixes.

I have also added another Cop to combine multiple calls with different shells into one.

The only thing I haven't been able to figure out is how to instruct the corrector to also remove the empty lines left behind, as can be seen in the homebrew/core example PR. I have looked into the `source_range` argument without success, but maybe @Rylan12 can give me a hint 😄

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -464,38 +464,45 @@ class GenerateCompletionsDSL < FormulaCop
-            binary = binary.source
+            shell = shell.to_s.delete_suffix("_completion").to_sym
+            executable = executable.source
             shell_parameter_stripped = shell_parameter.sub("bash", "").sub("zsh", "").sub("fish", "")
```
Does this need to be only stripped at the end?

Maybe

```suggestion
            shell_parameter_stripped = shell_parameter.chomp("bash").chomp("zsh").chomp("fish")
```

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -464,38 +464,45 @@ class GenerateCompletionsDSL < FormulaCop
-              :flag
+              ":flag"
+            elsif shell_parameter_stripped == "--shell="
+              ":arg"
```
FYI you could probably just leave this as a symbol and then use `#inspect` later to put the `:` and/or `"`

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -464,38 +464,45 @@ class GenerateCompletionsDSL < FormulaCop
             offending_node(node)
-            replacement = "generate_completions(#{replacement_args.join(", ")})"
+            replacement = "generate_completions_from_executable"
+            replacement += "(#{replacement_args.join(", ")})" unless replacement_args.blank?
```
`replacement_args` will never be empty because we always add the `shells` parameter, right?

max-ae replied:
You are totally right, this was left over from an attempt where I tried to combine all shells into a single call / Cop, when there was also a condition on the `shells` line.

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -517,14 +524,64 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+            offending_node(node)
+            problem "Use a single `generate_completions_from_executable` call
+                         combining all specified shells" do |corrector|
+              corrector.replace(@offensive_node.source_range, "")
```
Not sure this will work, but how about extending the replacement range by 1 extra character to include the `\n`, and then replacing that all with `""`?

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -517,14 +524,64 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          offending_node(offenses.last)
+          replacement = if (shells - %w[:bash :zsh :fish]).empty?
+            @offensive_node.source.sub(/shells: \[(:bash|:zsh|:fish)\]/, "")
+                           .gsub(",,", ",")
+                           .sub(", )", ")")
+                           .sub("(, ", "(")
+                           .sub("()", "")
```
What do all of these do?

max-ae replied:
Clean up the replacement string after updating the `shells` argument (because I did not figure these cases out in the regex, and I don't think that's even possible):
1. might not be needed, on second thought
2. `shells` was the last argument, the substitution would leave a trailing comma behind
3. `shells` was the first argument, the substitution would leave a leading comma in the front
4. `shells` was the only argument, the substitution would leave empty parentheses behind

Rylan12 replied:
What about something like this:

```ruby
            @offensive_node.source.sub(/(,\s*)?shells: \[(:bash|:zsh|:fish)\]/, "").sub(/,\s*\)/, "")
```

That will include the comma before, and then we only have to deal with the case of the comma after.

I'm pretty sure there's already a rubocop for empty parentheses that will take care of that for you (would be good to double check, though)

You could also use a single regex with `(,\s*)?` on both ends and then do some more manual parsing, maybe by passing a block to `sub`. Not 100% sure. I think a comment here would be useful if you want to stick with this, though

max-ae replied:
> ```ruby
>             @offensive_node.source.sub(/(,\s*)?shells: \[(:bash|:zsh|:fish)\]/, "").sub(/,\s*\)/, "")
> ```

This did not quite work. But you were right on the cop cleaning up parentheses!
I've removed the unnecessary cases and added comments for the remaining two.

Rylan12 reviewed (commented):
Looking good! I did leave a comment with a suggestion for the newline issue

max-ae commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -517,14 +524,64 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          end
+
+          offending_node(offenses.last)
+          replacement = if (shells - %w[:bash :zsh :fish]).empty?
```
I think this needs to be the other way round, now that I look at it again.

max-ae commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -455,6 +455,135 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+            end
+
+            replacement_args = %w[]
+            replacement_args << executable
```
No, since this is already the pathname in the form `bin/"foo"`. Calling `#inspect` would result in `"bin/\"foo\""`.

Rylan12 replied:
Does it work like this? I'd have thought you'd need to capture the entire node and then call `#source` to get that to work

max-ae replied:
It does, I suppose since it is the full Pathname node.
I figure the same procedure without `#inspect` might be working for the subcommand too, if I change the node matcher to capture the whole `str` node like this `$(str _)` instead of `(str $_)`.

Rylan12 replied:
Ah, sorry, I missed the `executable = executable.source` line above. Since we don't use `executable` anywhere else I'd suggest just removing that and calling `executable.source` on this line.

max-ae replied:
Yeah, you're totally right!

max-ae commented on Library/Homebrew/test/rubocops/lines_spec.rb:
```diff
@@ -58,3 +58,75 @@ class Foo < Formula
+        def install
+          (bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
+          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Use `generate_completions_from_executable(bin/"foo", "completions", shells: [:bash])` instead of `(bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")`.
+        end
```
This test fails at the moment, as the RuboCop expects an offense including `base_name` in the correction.
It seems like `@formula_name` is not instantiated correctly, since we are loading the formula from a heredoc and not a file path. Therefore, the `unless base_name == @formula_name` assertion in the Cop fails, and includes the `base_name` in the replacement.
Any suggestions how to solve this? I tried setting `cop.file_path= "foo.fb", but this did not help.
Should I include the `base_name` in the expected offense to make the test pass, even if this is not the wanted behavior?

Rylan12 replied:
Try changing the `expect_offense` line to:

```ruby
expect_offense(<<~RUBY, "/homebrew-core/Formula/foo.rb")
  class Foo < Formula
  ...
RUBY
```

max-ae commented:
> Can the tests be moved to their own file in `test/rubocops/lines/generate_completions_spec.rb`?

Sure, done!

Rylan12 commented on Library/Homebrew/test/rubocops/lines/generate_completions_spec.rb:
```diff
@@ -0,0 +1,78 @@
+
+require "rubocops/lines"
+
+describe RuboCop::Cop::FormulaAudit do
```
I don't think this top-level `describe` is needed

max-ae replied:
Without: 

```
Library/Homebrew/test/rubocops/lines/generate_completions_spec.rb:6:1: C: Do not use multiple top-level example groups - try to nest them.
describe RuboCop::Cop::FormulaAudit::GenerateCompletionsDSL do
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Rylan12 commented on Library/Homebrew/test/rubocops/lines/generate_completions_spec.rb:
```diff
@@ -0,0 +1,78 @@
+          name "foo"
+
+          def install
+            (bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
```
It would be nice to test some of the other features in this out. For example, `base_name`, different shell flags, etc.

max-ae replied:
Sure, I will add these cases as soon as I find time the next few days!

max-ae commented:
I have now extended the tests. 😄

Now that https://github.com/Homebrew/homebrew-core/pull/105707 is also merged, this should be good to go if CI passes 🎉

max-ae commented:
Yeah, I'm gonna need some help with that CI error...
I feel like I am missing a `return unless` somewhere?

```
2022-09-05T17:58:39.3127476Z [31mAn error occurred while FormulaAudit/GenerateCompletionsDSL cop was inspecting /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-portable-ruby/Abstract/portable-formula.rb:72:0.[0m
2022-09-05T17:58:39.3130785Z To see the complete backtrace run rubocop -d.
2022-09-05T17:58:39.3134252Z [31mAn error occurred while FormulaAudit/SingleGenerateCompletionsDSLCall cop was inspecting /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-portable-ruby/Abstract/portable-formula.rb:72:0.[0m
2022-09-05T17:58:39.3137255Z To see the complete backtrace run rubocop -d.
2022-09-05T17:58:39.6834742Z [31mAn error occurred while FormulaAudit/GenerateCompletionsDSL cop was inspecting /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-portable-ruby/Abstract/portable-formula.rb:72:0.[0m
2022-09-05T17:58:39.6835363Z To see the complete backtrace run rubocop -d.
2022-09-05T17:58:39.6836495Z [31mAn error occurred while FormulaAudit/SingleGenerateCompletionsDSLCall cop was inspecting /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-portable-ruby/Abstract/portable-formula.rb:72:0.[0m
2022-09-05T17:58:39.6838375Z To see the complete backtrace run rubocop -d.
2022-09-05T17:58:40.0134391Z [31m
2022-09-05T17:58:40.0135188Z 2 errors occurred:[0m
2022-09-05T17:58:40.0136425Z [31mAn error occurred while FormulaAudit/GenerateCompletionsDSL cop was inspecting /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-portable-ruby/Abstract/portable-formula.rb:72:0.[0m
2022-09-05T17:58:40.0137372Z [31mAn error occurred while FormulaAudit/SingleGenerateCompletionsDSLCall cop was inspecting /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-portable-ruby/Abstract/portable-formula.rb:72:0.[0m
2022-09-05T17:58:40.0154164Z undefined method `metadata' for nil:NilClass
2022-09-05T17:58:40.0156644Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command/execute_runner.rb:79:in `display_error_summary'
2022-09-05T17:58:40.0157372Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command/execute_runner.rb:58:in `display_summary'
2022-09-05T17:58:40.0158062Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command/execute_runner.rb:27:in `block in execute_runner'
2022-09-05T17:58:40.0158740Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
2022-09-05T17:58:40.0159393Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
2022-09-05T17:58:40.0160202Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
2022-09-05T17:58:40.0160832Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/command.rb:11:in `run'
2022-09-05T17:58:40.0161427Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli/environment.rb:18:in `run'
2022-09-05T17:58:40.0162020Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli.rb:72:in `run_command'
2022-09-05T17:58:40.0162602Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli.rb:79:in `execute_runners'
2022-09-05T17:58:40.0163177Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/rubocop-1.35.1/lib/rubocop/cli.rb:48:in `run'
2022-09-05T17:58:40.0163588Z /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/utils/rubocop.rb:12:in `<main>'

```

I can reproduce the failure if I have a formula containing only an `install` def, but I don't know why the cop should fail then... We are not calling `#metadata` anywhere...

E.g.
```ruby
class Foo < Formula
  def install
    (bash_completion/"foo").write Utils.safe_popen_read(bin/"foo", "completions", "bash")
  end
end
```

max-ae commented:
Also, the cask audit test failures should be unrelated.

max-ae commented:
Or: should we simply scope this cop to the core tap? The run for all core formulae was fine.

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -455,6 +455,149 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+        extend AutoCorrector
+
+        def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          install = find_method_def(body_node, :install)
```
This can be `nil` if there's no install method (which could happen with abstract formulae as we do in `Homebrew/homebrew-portable-ruby`. That seems to be what's causing the problem. You can fix this by adding

```suggestion
          install = find_method_def(body_node, :install)
          return if install.blank?
```

max-ae replied:
Thanks, that's what I was missing

Rylan12 commented on Library/Homebrew/rubocops/lines.rb:
```diff
@@ -455,6 +455,149 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+        extend AutoCorrector
+
+        def audit_formula(_node, _class_node, _parent_class_node, body_node)
+          install = find_method_def(body_node, :install)
```
Same here

```suggestion
          install = find_method_def(body_node, :install)
          return if install.blank?
```

Rylan12 reviewed (commented):
I see those audit test failures even on `master` locally. Not sure what's going on with them and why they haven't been showing up. I guess I'll look into it a bit later today.

max-ae commented:
Yay, all green now! (except for the unrelated tests) Thanks again!

That should be everything now for the new DSL 🎉

Rylan12 commented:
@SMillerDev the current CI issues (which are occurring for me locally even on the `master` branch) seem to caused by or related to https://github.com/Homebrew/brew/pull/13746. Any ideas?

Weirdly, I can see the errors if I run `brew tests --only=cask/audit` but if I add a line number e.g. `brew tests --only=cask/audit:6` (which should just run all the tests) I don't see an error.

SMillerDev commented:
I have no idea, it passed for me and in CI.

carlocab commented:
Failures likely fixed by #13813.

max-ae commented:
> Failures likely fixed by #13813.

Curiously, they are still present 🧐
Do I need to rebase?

carlocab commented:
> Do I need to rebase?

Yes, probably.

Rylan12 commented:
Thanks so much for all of your work on this, @max-ae! 🎉
